modules/resource_hwloc: new hwloc resource grokker

trws opened the pull request:
Currently in for discussion, not ready for general use as yet.  When loaded on
all ranks, `flux module load -r all resource_hwloc`, and loaded, `flux event
pub resource.load`, populates the KVS with resource information for the flux
instance.  The full topology for each node is stored as xml at
`resource.hwloc.xml.<rank>` in the KVS, and a complete tree of resources,
including IO and caches, by individual key and value is available at
`resource.hwloc.<rank>.[<resource_type>_<resource_logical_id>...]`. It has not
been instructed to look outside of administrative limitations, but that hasn't
been tested either.

trws commented:
This is mostly here for discussion right now.  As a test of the communication and aggregation stuff, this is an lstopo for 120 nodes of hype [topo-10.pdf](https://github.com/flux-framework/flux-core/files/1134/topo-10.pdf).

Script to print an xml topology suitable for lstopo for all ranks of the current instance:

```python
import flux
print Flux().rpc_send('resource.topo')['topology']
```

grondo commented:
That's pretty cool! I resubmitted the clang test, it failed due to intermittent wreckrun hangs we've been seeing.

For consistency with naming convention established by connector modules, this should probably be `resource-hwloc`.

So a basic scheduler implementation built on this could issue `resource.load` event, wait for completion, then walk the kvs entries under `resource.hwloc` and create a `node` object for each `rank` directory, and add available resources by traversing directories under the node's rank dir?

That is pretty nice.

trws commented:
Thanks! 

It still needs some cleanup, and the interface is what I think needs the most thought still.  Letting users walk the KVS directory is something I want to support, but I think it should directly provide a basic query interface as well, a-la #294, just "get #<type" and "<type> with at least <n> of <type> as children" that kind of thing.

Part of the reason for that would be to make it inter-operable with the RDL, but more importantly it would be a single central place to deal with the notion of what a "core" is to flux, and also deal with the inherent variability in system structures in the data.  It also means some of the hwloc query functionality can be leveraged, which will make some of this faster/easier than it would be to walk it all out of the KVS.

Thanks for the assist with the clang tests btw, I seem to keep running into some of those transient failures.

grondo commented:
Agreed on all counts, let me know if/how I can help. This will be a cool step forward. 

On the tests, there is a definite bug in wreckrun causing hangs, so hopefully we can replace or fix that interface soon

grondo commented:
For the query interface, do you envision each resource provider (e.g. `resource-hwloc, resource-sysfs, resource-config ... `, etc) providing the interface directly, or will these resource providers instead provide resources in a common form, which is then used to initialize RDL or other internal representation which in turn is the provider of the query interface?

I was leaning towards the latter, but I'm honestly not sure if that is the right approach or not.

In any event, it will be nice to have more than one `resource` providers loaded, but we've discussed this idea fully in #293 and related issues. However, in this case it might be useful to have access to *both* the "configured" resources for an instance (which may include things that cannot be discovered by hwloc -- licenses for example), and the hwloc topology, which may have useful information not in the configuration (like cache locality, pci devices, etc.). This is why I lean towards the ability to load multiple resource-* providers without them overriding each others methods, then using one representation to initialize the underlying provider of resource query api, which would be used by tools and scheduler.

Also, we'll need to decide how resources assigned to a program (for the near-term milestone in-job scheduler at least) get communicated to the parallel launch facility (currently the `wreck` prototype uses a kludgy approach of noting number of tasks to launch under `lwj.<id>.rank.<nodeid>`). We'll also want to be able to track which resources are in-use vs free (at least for the scheduler, but it would be nice to support other consumers of this data so we can get a top-like utility out of it)

(sorry, maybe a couple of these need to be full-blown issues, feel free to open any you see fit)

lipari commented:
I wonder whether we would find a faux resource reporting service to be of any use.  This would involve creating multiple names for the same resource, and allow to emulate more hardware than we actually have.  While this ability has served us well in prior RM's, it may (or may not) be useful in Flux.

trws commented:
@lipari: Once we figure out how the interface is to work, having a mock version would be a good idea if only for testing other components that depend on it.  That could also be the standard way to inject arbitrary resource types I suppose.

@grondo: I lean toward having a single interface, but allowing specific versions, or maybe a composite version, to be queried with that interface.  As you say, having more than one loaded would be very helpful, and having just a single endpoint that didn't do aggregation would prevent that.  It could be that we have one `resources` target that agregates results from the others, or each could provide the interface both at `resources` and `resources.<provider>` at the same time.  I had been planning to do a test implementation of the latter off of this PR, providing an interface at both names, then if an aggregation interface is desired it could be written against those secondary names.

As to dynamic properties of resources, I'm thinking that should be something that is layered on top of things like the hwloc resource interface.  Either as part of something like the RDL or as part of the scheduler.  The resource passing I hadn't really dealt with, other than to have a way to get the format that wreck expects to be written to the KVS.  Perhaps that needs its own ticket... The on-the-wire resource representation should really be standardized, just as we discussed for a basic jobspec in another ticket.

lipari commented:
I can see two ways to go with regard to creating the get and query interfaces of the type envisioned in #294.  One option is to support complex queries that navigates the copious info currently generated by resource_hwloc to find the items of interest:  sockets, cores, L3 cache, etc.  The other option is to modify resource_hwloc to only report/record the items of interest.  The latter option would conceivably simplify the queries but limit the scope of what clients would see.

garlick commented:
Would it make sense to get the raw "populate kvs" part cleaned up and finalized and then do the get/query interfaces as a new PR?

trws commented:
I think the answer to this is "yes," I just have to get my head back 
above water far enough to finish this pull request off.  It's actually 
in a semi-reasonable state now, but I feel like it should be wired up to 
load by default rather than having to be manually loaded and invoked.

On 17 Aug 2015, at 19:23, Jim Garlick wrote:

> Would it make sense to get the raw "populate kvs" part cleaned up and 
> finalized and then do the get/query interfaces as a new PR?
>
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/flux-framework/flux-core/pull/312#issuecomment-132039580

garlick commented:
The way to do that is (embarassingly) edit the static string `default_modules` in `src/broker/broker.c`.

Are we OK with this module's behavior of loading the entire node's topo under each rank, even when multiple ranks per node are launched (e.g. in test?)

trws commented:
Nope. I had a note to do de-duplication based on hostname (the hwloc info actually includes it), but hadn't decided whether to do that by explicitly checking or just having each of them write to a path prefixed by its hostname and let the KVS figure it out.  What do you think?

grondo commented:
Warning: Opinion-heavy piece follows:

I makes a kind of sense to leave the duplication there since there may actually be some reasonably shared topology on which multiple brokers are running (e.g., network switches in the future), and the "real" hardware topology store ought to reflect that. (another e.g., if running one broker per "socket" you wouldn't want to/couldn't dedup on nodes)

For test purposes, we should allow seeding hwloc information with "fake" topology data (I believe you can load in an xml description and use that instead of "real" hardware) to test various scenarios that use the hardware information, but most of the stuff under `./t` probably doesn't care at this point.

Other times when running multiple brokers per node `flux start` as bootstrap instance could/should contain brokers to resources (e.g. divide cores among brokers, perhaps there could be a plugin for this) possibly just using cpu_affinity for now, and the hwloc stuff *should* just report hardware to which the broker in which it is loaded has access.

garlick commented:
Would it make sense then for this initial PR to store the topo under the hostname as @trws suggests, letting the KVS squash the duplicates, and then store a symlink to the hostname under the rank?  That would capture the full picture.

trws commented:
Good points.  I had been thinking that doing KVS-based de-dup would be 
reasonable because the other hardware in the tree would still be using 
the OS IDs, and thus they would all merge into a single representative 
hierarchy, but then it would be impossible to separate out the part 
belonging to just that rank.

I need to test it, but in principle the flags I gave hwloc should only 
show those resources visible to the process making the request.  It's 
actually optional, I could have it list all resources but mark those 
that are unavailable as such if we prefer that.

On 17 Aug 2015, at 20:29, Mark Grondona wrote:

> Warning: Opinion-heavy piece follows:
>
> I makes a kind of sense to leave the duplication there since there may 
> actually be some reasonably shared topology on which multiple brokers 
> are running (e.g., network switches in the future), and the "real" 
> hardware topology store ought to reflect that. (another e.g., if 
> running one broker per "socket" you wouldn't want to/couldn't dedup on 
> nodes)
>
> For test purposes, we should allow seeding hwloc information with 
> "fake" topology data (I believe you can load in an xml description and 
> use that instead of "real" hardware) to test various scenarios that 
> use the hardware information, but most of the stuff under `./t` 
> probably doesn't care at this point.
>
> Other times when running multiple brokers per node `flux start` as 
> bootstrap instance could/should contain brokers to resources (e.g. 
> divide cores among brokers, perhaps there could be a plugin for this) 
> possibly just using cpu_affinity for now, and the hwloc stuff *should* 
> just report hardware to which the broker in which it is loaded has 
> access.
>
>
>
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/flux-framework/flux-core/pull/312#issuecomment-132057892

grondo commented:
> Would it make sense then for this initial PR to store the topo under the hostname as @trws suggests, letting the KVS squash the duplicates, and then store a symlink to the hostname under the rank? That would capture the full picture.

That might be a good short term solution but wouldn't work for the case we're discussing where each rank, say, gets only 1 of N cores on the node...

I guess we shouldn't optimize for the single node/test case. Storing duplicates when running multiple brokers on a single node probably isn't a big deal.

grondo commented:
Oops sorry @trws just noticed you said essentially the same thing.

trws commented:
By the by, you're right about the XML loading @grondo.  That's what allows it to generate an allocation-wide file for lstopo.  For implementing the request interface, I'm seriously thinking about setting it up to just do the queries against the complete aggregate since hwloc actually implements "minimum coverage" operators that actually handle almost all the basic cases we care about.

trws commented:
I'm leaning towards loading the data into the KVS this way.  Have one directory that contains a version grouped by hostname, hopefully creating a full picture of all the hardware available to the instance, and a second directory where the same information is grouped by rank.  That way we can differentiate between what exists and what ranks are assigned.  It may also be worthwhile to "tag" each item in the hostname based hierarchy with the ranks that see it.   I realize this means more data, but it seems more overall useful to me.

trws commented:
The new version just pushed uses the scheme I described earlier to store information, and also stores the number of each type of object available to a rank in an int of the type name in the rank's directory.  This should be more or less good to go, short only the resource request interface as discussed earlier this afternoon.

garlick commented:
It seems to me that the three patches that add and then modify the module should be squashed, and the commit message should be revised to drop the for review only language and to conform to the new patch requirements in  [rfc 1](https://github.com/flux-framework/rfc/blob/master/spec_1.adoc).

If you have time, you might try converting this to the new reactor interface also.  Not a big deal as I'm sure I'll be doing a big PR to convert the whole code base in the near future, but your feedback would be valuable.

trws commented:
Good points.  I've got it rebased, and an converting the reactor interface now.  Did anything ever happen with getting `__attribute__ ((deprecated))` on interfaces that we're planning to phase out? I didn't even notice I was accidentally using the older one.

garlick commented:
As I recall the problem with the deprecated attribute is then that we have to disable `-Werror` because there are still many users of the deprecated functions in tree.

trws commented:
Ah... That's unfortunate. I wonder if there's a way to make an exception to Werror for a particular warning type.

grondo commented:
For GCC
```
       -Werror=
           Make the specified warning into an error.  The specifier for a
           warning is appended, for example -Werror=switch turns the warnings
           controlled by -Wswitch into errors.  This switch takes a negative
           form, to be used to negate -Werror for specific warnings, for
           example -Wno-error=switch makes -Wswitch warnings not be errors,
           even when -Werror is in effect.  You can use the
           -fdiagnostics-show-option option to have each controllable warning
           amended with the option which controls it, to determine what to use
           with this option.
```

trws commented:
Thanks @grondo, so `-Werror -Wno-error=deprecated-declarations` should do it right?

garlick commented:
Oh nice, that works.  I'll submit a PR shortly.  Thank you guys!

trws commented:
Now that travis passed, take a look at the updated version @garlick, I think it covers updating to the new reactor interface.  Since these are all message watchers it doesn't really come up here, but working on the submission tool (having some trouble getting the scheduler to do my bidding... issues coming) I'm noticing that I desire the different watchers to be polymorphic more than I had previously.  Having the different start/stop/destroy interface for each one is a bit unfortunate.  Otherwise I find it pretty intuitive.

garlick commented:
I'm off to lunch but will review when I get back.

garlick commented:
Shouldn't the module name be declared as "resource_hwloc" not "resource"?  Or were we expecting only one resource module to be loaded at a time?

Maybe you chose not to use it (and that's perfectly fine) but FYI there is a shorthand function for initializing multiple message watchers: `flux_msg_watcher_addvec()`.

Your comment on unfortunate lack of polymorphism is noted - in #219 we decided not to do this, but if we'd like to reopen the discussion, let's get a new issue open.

Your `struct flux_match` literal argument caused me to go verify that those are handled like static initializers, so the unmentioned arguments are zeroed per C99.  Also, I note that the preprocessor macros FLUX_MATCH_ANY and FLUX_MATCH_EVENT could be used as arguments if they included the cast.  (I was irritated that this didn't work but didn't think of casting! duh)

Looks fine to me otherwise.  Maybe @grondo should give an ack on the way data is stored.

grondo commented:
Yeah, I need to take a peek at this though I'm sure it is fine.

After this goes in, I'd like to do whatever's needed to remove the barebones `resrc` module under wreck/ in the short term.

I also thought module would be `resource-hwloc` (i.e. to match precedent set by `connector-local`) I think we had some discussion on whether there would be more than one "provider" of `resource` service, and I thought we had decided yes, as you may want to provide more resource information than hwloc could provide alone (e.g. switches, licenses), but still might want input from hwloc itself.

I think @trws had some definite ideas on how this could be done, but I'm not sure I ever got the full picture

trws commented:
Yeah, the name of the module should include the dash, I'll adjust that shortly.  The addvec would probably be cleaner, as would using the MATCH_EVENT macro.

I have a vague recollection that the reason I had it be "resource" rather than "resource_hwloc" had to do with the interaction between the module name and the service name for requests.  Do we have a restriction there or is this something I'm pulling out of thin air?

garlick commented:
For now the service name must match the module name.  Also, the convention established by `connector-local` uses hyphen not underbar.

trws commented:
So it does @garlick, but your example from the other comment used underbar.  I'll switch this over since the interface is being delayed, but the reasoning was that the interface should be available at `resource` rather than at `resource-*`, and the only way to do that was to have the module also named `resource`.

grondo commented:
Yeah, there is the whole discussion of how we will create new services that "provide" or "override" a standard flux service, like "resource" in this instance.

The "resource" service might provide a standard, perhaps extensible, resource status and query API (or perhaps kvs-based schema), but ideally you would build this service on top of one more resource data providers like resource-hwloc. (However, resource-hwloc itself would not be able to provide *all* resource data, as it cannot handle anything that exists outside OS purview)

I thought we had a good discussion of this going previously, though I'm not seeing it now.

garlick commented:
I think #293 contains the discussion?

Sorry @trws, both for the gap between reality and where we'd like to be, and for the underbar/hyphen shenanigans!

grondo commented:
Actually I was thinking more specifically how a `resource` service would be built from various resource data providers like `resource-hwloc` and perhaps `rdl`.

For me it seems most intuitive that `resource-hwloc` and (for example) `resource-conf` are resource configuration providers and would act like backend initial configuration for a standard `resource` interface to which the scheduler and users interact.

However, I had thought @trws at some point had some other clever ideas about how to tie multiple resource providers into a single service... not sure now though.

trws commented:
My thought process had been that each of them would provide an interface called `resource` and act as a hierarchy or chain of responders.  The hwloc version might be the bottom default one, with each subsequent resource provider building on top of it, and if they didn't have the answer passing the message along to the next handler in the chain.  That way the composition of the different providers happens as a direct result of their arrangement, rather than through a separate module that explicitly queries each one.  Hence the original desire to have this provide `resource` as a service, then with the addition of a "send request to next implementer of this name" functionality, any number of others could be layered on top using the same protocol.

grondo commented:
@trws, that sounds reasonable to me -- have we captured the need for scoped services in some Issue somewhere?

trws commented:
It was under issue #293, if we had that capability that would be all we 
need for a singly-linked hierarchy like I was describing above.

On 20 Aug 2015, at 15:03, Mark Grondona wrote:

> @trws, that sounds reasonable to me -- have we captured the need for 
> scoped services in some Issue somewhere?
>
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/flux-framework/flux-core/pull/312#issuecomment-133192081

trws commented:
Actually... that talks about overriding, I'm not sure about the module name -> service name mapping issue, which is also important for this so we don't have modules clobbering one another.

grondo commented:
This looks great to me. For `flux-wreckrun` `alloc_tasks_hack` I'd like to adopt usage of `resource.hwloc` information instead of `resrc`, then we can delete the `resrc` module (I think that is the only place we use `resrc` for now but I'll double check)

I think all flux-wreckrun needs is the number of cores per rank and it appears this is available as 
```
resource.hwloc.by_rank.<id>.Core
```

trws commented:
Yup, that's where it is.

grondo commented:
One thing I noticed is that even running across multiple physical nodes, the "Machine" is always `Machine_0`. I imagine this is expected because we're running libhwloc singleton on each "machine", however, it would be nice if there were a way to easily differentiate the machines on which each rank is running? Perhaps even a `by_rank.<id>.hostname` or other entry in the by_rank summary?

Actually, I wonder if the `by_rank.<id>.hostname` could just be a link the the `by_host.<hostname>` directory? (I know we already discussed this to death here but I couldn't help bringing it up)

Those issues could be discussed later and this PR could be merged I think (nice feature!), once the module name becomes `resource-hwloc` (and we should probably make the directory name match as well, it is currently `resource_hwloc`)

trws commented:
Yeah, they all show up as Machine_0 because it's the root machine of that node, and the start of the tree walk.  That said, they all have `by_rank.<id>.HostName` in them to disambiguate if you want to, and that value can be used to find the host in the `by_host` directory.  We could also add a symlink though if that's helpful.

I'll get the names fixed up as soon as I push in an initial version of this (very, very prototype) bulk run tool.

grondo commented:
I didn't see the `HostName` entry, which would be perfect. (Actually realizing now that the link is a bad idea).

```$ src/cmd/flux kvs dir resource.hwloc.by_rank.1          
resource.hwloc.by_rank.1.Machine = 1
resource.hwloc.by_rank.1.NUMANode = 2
resource.hwloc.by_rank.1.Socket = 2
resource.hwloc.by_rank.1.Cache = 16
resource.hwloc.by_rank.1.Core = 16
resource.hwloc.by_rank.1.PU = 16
resource.hwloc.by_rank.1.Machine_0.
```

Ah, now I see it is under `by_rank.<id>.Machine_0.HostName`:

```
$ src/cmd/flux kvs dir resource.hwloc.by_rank.1.Machine_0 | grep HostName
resource.hwloc.by_rank.1.Machine_0.HostName = hype216
```

Are we thinking there may be other `Machine`'s added to ranks in the future and that is why there is a numbered subdir there?

Anyhow, this works for me. Thanks!

trws commented:
It was mainly just to make it uniform, I set up the module to pull all attributes out of all objects the same way, including the machine, and this is what came out (hence the camel-case name "HostName" too...).  That said, we actually might have more than one machine in a single topology at some point, if you use the `resource.topo` request to get an overall topology, it will give you back an xml topology containing all of the machines in the job as sub-objects.  I'm not sure if we'll ever want to dump one of those straight into the KVS, but it might be handy for a parent instance to be able to do I suppose.

grondo commented:
That seems reasonable -- and a good argument for not promoting hostname up to the by_rank.<id> level.

btw, is there a `flux lstopo` tool that goes with this module ;-)

trws commented:
Actually, I have a python script that implements that... I can pop that in on the pull request if you want it. =)

grondo commented:
Well if nothing else it makes a cool demo. We can get the base support in first though.

trws commented:
Ok, it's renamed, moved and makes better use of the `FLUX_MATCH*` macros.  The `flux-lstopo` script is now in its own PR over at #345.

grondo commented:
This should get merged once ef337ec and 2f27bc5 are squashed together (no reason to persist name change in ultimate history) Thanks!

trws commented:
Squashed and rebased.  Oddly github seems to be showing the last two commits out of order, hopefully that won't cause problems with travis, but if it does that's all it is.

grondo commented:
I've noticed that with github recently -- it sometimes seems to do some weird sorting on commits in a PR (very confusing at times!)

grondo commented:
Great! Merging...
